picard/browser/browser.py

```diff
@@ -25,115 +25,169 @@
-                    parsed = urlparse(line[1])
-                    args = parse_qs(parsed.query)
-                    if 'id' in args and args['id']:
-                        mbid = args['id'][0]
-                        if not mbid_validate(mbid):
-                            log.error("Browser integration failed: bad mbid %r", mbid)
-                            return False
-
-                        def load_it(loader):
-                            self.tagger.bring_tagger_front()
-                            loader(mbid)
-                            return True
-                        action = parsed.path
-                        if action == '/openalbum':
-                            return load_it(self.tagger.load_album)
-                        elif action == '/opennat':
-                            return load_it(self.tagger.load_nat)
-            except Exception as e:
-                log.error("Browser integration failed with %r on line %r", e, orig_line)
-                return False
-            log.error("Browser integration failed: cannot parse %r", orig_line)
-            return False
+class RequestHandler(BaseHTTPRequestHandler):
+
+    def do_OPTIONS(self):
+        origin = self.headers['origin']
+        if _is_valid_origin(origin):
+            self.send_response(204)
+            self.send_header('Access-Control-Allow-Origin', origin)
+            self.send_header('Access-Control-Allow-Methods', 'GET')
+            self.send_header('Access-Control-Allow-Credentials', 'false')
+            self.send_header('Access-Control-Allow-Private-Network', 'true')
+            self.send_header('Access-Control-Max-Age', 3600)
+            self.send_header('Vary', 'Origin')
+            self.end_headers()
+        else:
+            self.send_response(401)
+            self.end_headers()
 
+    def do_GET(self):
         try:
-            line = rawline.decode()
-            if parse_line(line):
-                conn.write(response(200))
+            self._handle_get()
+        except Exception as e:
+            self._response(500, str(e))
+
+    def _handle_get(self):
+        parsed = urlparse(self.path)
+        args = parse_qs(parsed.query)
+        action = parsed.path
+
+        if action == '/':
+            self._response(200, SERVER_VERSION)
+        elif 'id' in args and args['id']:
+            mbid = args['id'][0]
+            if self._load_mbid(action, mbid):
+                self._response(200, _('MBID "%s" loaded') % mbid)
             else:
-                conn.write(response(400))
-        except UnicodeDecodeError as e:
-            conn.write(response(500))
-            log.error(e)
-            return
-        finally:
-            conn.disconnectFromHost()
-
-    def _accept_connection(self):
-        conn = self.nextPendingConnection()
-        conn.readyRead.connect(self._process_request)
+                self._response(400, _('Could not load MBID "%s"') % mbid)
```

Comment:
I would avoid returning a non-validated string here, just something like "id isn't a valid mbid".
Also it makes no sense to translate those strings imho (client may use a different language).